src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultApplicationModelProvider.cs

```diff
@@ -572,6 +589,19 @@ private IList<SelectorModel> CreateSelectors(IList<object> attributes)
             return selectorModels;
         }
 
+        private static bool InRouteProviders(List<IRouteTemplateProvider> routeProviders, object attribute)
+        {
+            foreach (var rp in routeProviders)
```

Comment:
Is perf your concern here? We are iterating over List<T> so no allocation. We do foreach elsewhere in the class if I recall correctly.